datadog_checks_base/datadog_checks/base/utils/db/statement_metrics.py

```diff
@@ -0,0 +1,127 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class StatementMetrics:
+    """
+    This class supports normalized statement-level metrics, which are collected from the database's
+    statistics tables, ex:
+
+        - Postgres: pg_stat_statements
+        - MySQL: performance_schema.events_statements_summary_by_digest
+        - Oracle: V$SQLAREA
+        - SQL Server: sys.dm_exec_query_stats
+        - DB2: mon_db_summary
+
+    These tables are monotonically increasing.
+    """
+
+    def __init__(self):
+        self.previous_statements = {}
+
+    def compute_derivative_rows(self, rows, metrics, key):
+        """
+        Compute the first derivative of column-based metrics for a given set of rows. This also resets the
+        statement cache so should only be called once per check run.
+
+        - **rows** (_List[dict]_) - rows from current check run
+        - **metrics** (_List[str]_) - the metrics to compute for each row
+        - **key** (_callable_) - function for an ID which uniquely identifies a row across runs
+        """
+        result = []
+        new_cache = {}
+        negative_result_found = False
+        metrics = set(metrics)
+        if len(rows) > 0:
+            dropped_metrics = metrics - set(rows[0].keys())
+            if dropped_metrics:
+                logger.warning(
+                    'Some statement metrics are not available from the table: %s', ','.join(m for m in dropped_metrics)
+                )
+
+        for row in rows:
+            row_key = key(row)
+            if row_key in new_cache:
+                logger.debug(
+                    'Collision in cached query metrics. Dropping existing row, row_key=%s new=%s dropped=%s',
+                    row_key,
+                    row,
+                    new_cache[row_key],
+                )
+            new_cache[row_key] = row
+            prev = self.previous_statements.get(row_key)
+            if prev is None:
+                continue
+            metric_columns = metrics & set(row.keys())
+            if any([row[k] - prev[k] < 0 for k in metric_columns]):
+                # The table was truncated or stats reset; begin tracking again from this point
+                negative_result_found = True
+                continue
+            if all([row[k] - prev[k] == 0 for k in metric_columns]):
+                # No metrics to report; query did not run
+                continue
+            derived = {k: row[k] - prev[k] if k in metric_columns else row[k] for k in row.keys()}
+            result.append(derived)
+
+        self.previous_statements = new_cache
+        if negative_result_found:
+            return []
+        return result
+
+
+def apply_row_limits(rows, metric_limits, tiebreaker_metric, tiebreaker_reverse, key):
```

Comment:
I'm not sure I understand the motivation for this function, and what it does.

Why do we take the top `k_t` and bottom `k_b` rows (based on the metric value) each time? At first I thought the tuple meant `(<min_num_of_values>, <max_num_of_values>)`, without exactly knowing what that would mean either.

Maybe I'm just completely missing something, but could you give a run-through of what led to this logic?

There is a metric limits functionality built into integrations, which is simply "don't submit metrics once more than `N` contexts have been submitted" (with `N` being the configurable toggle). Logic can be found here: https://github.com/DataDog/integrations-core/blob/c667befb49ad8a2202174edb9e682c1eb773ac43/datadog_checks_base/datadog_checks/base/checks/base.py#L428-L438 and limiter is implemented here: https://github.com/DataDog/integrations-core/blob/1bf1bcf5f4af4a54f16bf45ad76d55c4f364ce05/datadog_checks_base/datadog_checks/base/utils/limiter.py

 Why do we need something more complex for database statement metrics?

Reply:
Sure, it might help to start with showing them in the product (see screenshot). Users will want to sort the table to show the "worst performing query by some metric." To make sure sorting is accurate, we cannot drop the true bottom or true top of any metric (it would be really bad if we said the slowest query is 500ms, but the slowest query is really 2 min). But we need to include the entire row for that query, even if the other metrics in the row are not in the top k or bottom k of their respective metric.

But you'll notice in that screenshot there are a lot of metrics "tied" for 0. The purpose of the "tiebreaker" is to aim for consistency in the ones we pick. The dict is for deduping.

If we just choose the top k or bottom k for each metric without regard for consistency, that table might look something like this since each metric is a separate column:

| count | time | shared_blks_hit  | shared_blks_dirtied | shared_blks_miss (sort asc) |
|-------|-----|-----------------|--------------------|------------------------------|
|            |        |                             |                                  | 0 |
|            |        |                             |                                  | 0 |
|            |        |  0                         |                                  | 0 |
|            |        |                             |                                  | 0 |
|  15       |        |                             |                                  | 0 |

This is because if we just took the bottom k of shared_blks_miss, then there is no guarantee the other metrics will be populated for that query's row.

<img width="1148" alt="Screen Shot 2020-10-26 at 9 48 08 AM" src="https://user-images.githubusercontent.com/717862/97180411-67ca0180-1770-11eb-896e-6bd8d6e6acc7.png">

Reply:
By the way, I'm not saying this is by any means the best way to solve this problem. I completely understand the confusion and I doubt this is what the implementation of limits will look like long-term. It is a bit subjective and I may have missed a much more obvious solution.

Reply:
@justiniso Yes, that's much clearer with some product background, thanks!